KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,238 @@
+- In case of a Contract, verification occurs when the admin of the contract calls approve function. The smart contract calls the `getState()` function implemented in the retiredAddress contract to get the admin details. `msg.sender == admin`
+  - `getState()` funtion is implemented in Klaytn treasury contracts. It returns the adminList and quorom (min required admins to approve).
+  - Condition: Min required admins should approve the retiredAddress contract.
+
+#### No Withdrawal
+
+Smart contract is not allowed to receive KLAY due to security reasons. So any transaction sending KLAY to the contract will be reverted and withdraw function is not implemented.
+
+#### Finalize Contract
+
+Once the re-distribution a.k.a rebalance is executed by the Core, the status of the smart contract will be finalized and any modifications to the storage data will be restricted.
+
+## Backwards Compatibility
+
+<!-- All KIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The KIP must explain how the author proposes to deal with these incompatibilities. KIP submissions without a sufficient backwards compatibility treatise may be rejected outright. The authors should answer the question: "Does this KIP require a hard fork?" -->
+
+- The foundation should deploy new TreasuryRebalance contract to record the token redistribution.
+- To rebalance the funds and redistribute in a consistent manner the foundation should burn the designated funds before re-distribution.
+- This does not affect the backward compatibility as this a newly dpeloyed contract
+
+## Implementation
+
+<!--The implementations must be completed before any KIP is given status "Final", but it need not be completed before the KIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+#### Example implementation
+
+```solidity
+interface ITreasuryRebalance {
+    // Events
+    event ContractDeployed(
+        uint256 rebalanceBlockNumber,
+        uint256 deployedBlockNumber
+    );
+    event RetiredRegistered(address retired);
+    event RetiredRemoved(address retired, uint256 retiredCount);
+    event NewbieRegistered(address newbie, uint256 fundAllocation);
+    event NewbieRemoved(address newbie, uint256 newbieCount);
+    event Approved(address retired, address approver, uint256 approversCount);
+    event StatusChanged(Status status);
+    event Finalized(string memo);
+
+    // Enums
+    enum Status {
+        Initialized,
+        Registered,
+        Approved,
+        Finalized
+    }
+
+    struct Retired {
+        address retired;
+        address[] approvers;
+    }
+
+    struct Newbie {
+        address newbie;
+        uint256 amount;
+    }
+
+    // State variables
```

Comment:
I assume `Retired[] retirees` and `Newbie[] newbies` are public state variables? Interfaces of them would be
```solidity
function retirees(uint256 index) external view returns (address);
function newbies(uint256 index) external view returns (address, uint256);
```
Note that `Retired.approvers` is not available from the automatic getters because it's a nested array.

If we want to either (1) get all retirees and newbies at one call, or (2) get `approvers` array, we need separate getters. For example,
```solidity
function getRetirees() external view returns (address[] memory, address[][] memory);
function getNewbies() external view returns (address[] memory, uint256[] memory);
```

Reply:
yes there are separate getters to get retired and newbie details

Reply:
If `finalizeContract()` is the getter you've mentioned, then I'm confused because `finalizeContract()` is apparently a state-changing function. Like @kjhman21 mentioned in the other comment, it might be easier to understand if this function is separated into e.g. a state-changing `finalizeContract()` and a getter `getStatus()`

Reply:
Now I see that `retirees(index)` and `getRetired(address)` combo can return every state somehow.
Then I am okay with the current interface. Can you just add `retirees` and `newbies` to ITreasuryRebalance for clarity?

Reply:
@toniya-klaytn `getSnapshot()` clears my concern. Thanks.